src/ploigos_step_runner/results/workflow_result.py

```diff
@@ -253,11 +248,39 @@ def write_to_pickle_file(self, pickle_filename):
         """
         try:
             create_parent_dir(pickle_filename)
+            # merge in steps that are in the on-disk pickle, but not in our
+            # in-memory copy.
+            on_disk_wfr = WorkflowResult.load_from_pickle_file(pickle_filename)
+            for step_result in on_disk_wfr.workflow_list:
```

Comment:
To merge on-disk WFR with in-memory WFR. Duplicate steps need to be skipped, otherwise we'll double up, so we go one SR at a time and skip the duplicates. Is there a helper function that already does this that I missed? It may be better for the logic to go into `add_step_result`, however it's not there currently.

Reply:
the problem i see now is if two parallel executions attempt to execute a step with the same name, sub step name, and environment, we will again be silently ignoring the results of one of the executions. though i dont have any briliant ideas for solving that, other then maybe as part of the "add_step_result" we mark the new step result as a "new one" somehow, and then when we do write it do disk we ditch that property so when it gets loaded in by another step it is no longer new. maybe we need to have two different arrays of step results in the workflow results, one of "old" and one of "new" and then when we write to disk, we merge the "new" over the to the "old" and then write, so then when the pickle gets loaded everything is in the "old" list, and then when doing the merge before the write, can check for duplicates and fail just as we always have. though this complicates a bunch more code, though it should all be limited to functions in WorkflowResults. Idk, any better ideas?

also i think the merge logic, shouldn't live in the write_to_pickle_file, it should live in a "reload_(and_merge?)_from_pickle" to have separation of concerns.

Reply:
Ok, I'll deal with both at once. I'll add the required merging logic into `add_step_result`, create a test for that logic, and refactor the merge code to another method.

Reply:
I've created `WorkflowResult.merge_from_pickle_file` that will merge and update the internal workflow list to include results on-disk. This uses the `StepResult.merge` method.

I've also added the `StepResult.merge` method that will merge artifacts and evidence from one `StepResult` into another, providing they have matching step name, sub-step name and environment. Artifact/evidence merger is straight replacement if artifact/evidence key already exists; I use `add_evidence` and `add_artifact` for this.